### PR TITLE
cgroupsgateway: add support for net_cls.classid

### PR DESCRIPTION
**Pull request by tobsan**
Signed-off-by: Tobias Olausson <tobias.olausson@pelagicore.com>

### CODE REVIEW

File: libsoftwarecontainer/src/gateway/cgroups/cgroupsparser.cpp [hunk 1]
```diff
@@ -100,18 +100,37 @@ void CGroupsParser::parseCGroupsGatewayConfiguration(const json_t *element)
...
+        if (settingValue.find("0x") != 0 // Has to begin with 0x
+         || settingValue.length() > 10 // Can not be longer than 0x + 8
+         || settingValue.length() < 7) // Has to be at least 0xAAAAB to cover major/minor handle
+        {
+            std::string errorMessage = "net_cls.classid should be of form 0xAAAABBBB";
+            log_error() << errorMessage;
+            throw HexFormatError(errorMessage);
+        }
+
+        char *endPtr;
```
Comment (DunderRoffe) [inline libsoftwarecontainer/src/gateway/cgroups/cgroupsparser.cpp]: This should be initialized.

**Review by DunderRoffe — APPROVED** [review]

File: libsoftwarecontainer/src/gateway/cgroups/cgroupsparser.cpp [hunk 2]
```diff
@@ -100,18 +99,37 @@ void CGroupsParser::parseCGroupsGatewayConfiguration(const json_t *element)
...
-                m_settings[settingKey] = settingValue;
+        // if the new value is smaller/equal than the old value, then we don't save the new value.
+        if (m_settings.count(settingKey) != 0) {
+            if (std::stoll(m_settings[settingKey]) >= std::stoll(settingValue)) {
+                return;
             }
-        } else {
-            m_settings[settingKey] = settingValue;
         }
+    } else if (("net_cls.classid") == settingKey) {
```
Comment (kursatkobya) [inline libsoftwarecontainer/src/gateway/cgroups/cgroupsparser.cpp]: removing parenthesis on  `("net_cls.classid") `  may be clearer to read here

File: libsoftwarecontainer/src/gateway/cgroups/cgroupsparser.cpp [hunk 3]
```diff
@@ -100,18 +99,37 @@ void CGroupsParser::parseCGroupsGatewayConfiguration(const json_t *element)
...
+         || settingValue.length() < 7) // Has to be at least 0xAAAAB to cover major/minor handle
+        {
+            std::string errorMessage = "net_cls.classid should be of form 0xAAAABBBB";
+            log_error() << errorMessage;
+            throw HexFormatError(errorMessage);
+        }
+
+        char *endPtr = NULL;
+        unsigned long int hexedValue = strtoul(settingValue.c_str(), &endPtr, 16);
+        if (hexedValue == 0 || hexedValue == ULONG_MAX || ((endPtr != NULL) && *endPtr != '\0')) {
```
Comment (kursatkobya) [inline libsoftwarecontainer/src/gateway/cgroups/cgroupsparser.cpp]: i would prefer `0 == hexedValue` and `ULONG_MAX == hexedValue` here

**Review by kursatkobya — APPROVED** [review]

File: doc/chapters/gateways/10-cgroups-gateway.rst [hunk 4]
```diff
@@ -38,6 +38,13 @@ memory, thus it cannot be less than ``memory.limit_in_bytes``.
 Also note that the one can use a suffix (k, K, m, M, g or G) to indicate values in kilo,
 mega or gigabytes when setting ``memory.limit_in_bytes`` or ``memory.memsw.limit_in_bytes``.
 
+Setting network classes
+-----------------------
+If you want to mark packets from containers, you can set the ``net_cls.classid`` to a hexadecimal
+value of the form 0xAAAABBBB (leading zeroes may be omitted for both A and B part). The gateway
+supports and checks the syntax for this key. More information about this can be found `in the Linux
+kernel docs <https://www.kernel.org/doc/Documentation/cgroup-v1/net_cls.txt>`_.
+
```
Comment (breakreturn) [inline doc/chapters/gateways/10-cgroups-gateway.rst]: I think this should be extended with an example. The example config section doesn't reflect this at all.

**Review by breakreturn — CHANGES_REQUESTED** [review]

**Review by kursatkobya — APPROVED** [review]
Also a service test would be nice

**Comment by tobsan** [comment]
Updated according to comments. Will add a service test also.

**Comment by tobsan** [comment]
Added a service test now.

File: servicetest/cgroups/test_cgroups.py [hunk 5]
```diff
@@ -54,40 +54,50 @@ def mounted_path_in_host():
     Container.READONLY: False
 }
 
-CONFIG_TEST_SMALL_THRESHOLD = [
-    {"setting": "memory.limit_in_bytes", "value": "1K"}
+CONFIG_TEST_MEMORY_SMALL_THRESHOLD = [
+    { "setting": "memory.limit_in_bytes", "value": "1K" }
```
Comment (breakreturn) [inline servicetest/cgroups/test_cgroups.py]: Spaces after the braces is not how the rest of the code looks, please align with the rest of the code or change everywhere.

**Review by breakreturn — CHANGES_REQUESTED** [review]

File: servicetest/cgroups/test_cgroups.py (hunk 5 above)
Comment (tobsan) [reply servicetest/cgroups/test_cgroups.py]: Sure, changed it back.

**Review by breakreturn — APPROVED** [review]

**Pull request merged**